p2p/transport/quic/quic_multiaddr.go

```diff
@@ -1,30 +1,62 @@
 package libp2pquic
 
 import (
+	"errors"
 	"net"
 
+	"github.com/lucas-clemente/quic-go"
 	ma "github.com/multiformats/go-multiaddr"
 	manet "github.com/multiformats/go-multiaddr/net"
 )
 
-var quicMA ma.Multiaddr
+var quicV1MA ma.Multiaddr = ma.StringCast("/quic-v1")
+var quicDraft29MA ma.Multiaddr = ma.StringCast("/quic")
 
-func init() {
-	var err error
-	quicMA, err = ma.NewMultiaddr("/quic")
-	if err != nil {
-		panic(err)
-	}
-}
-
-func toQuicMultiaddr(na net.Addr) (ma.Multiaddr, error) {
+func toQuicMultiaddr(na net.Addr, version quic.VersionNumber) (ma.Multiaddr, error) {
 	udpMA, err := manet.FromNetAddr(na)
 	if err != nil {
 		return nil, err
 	}
-	return udpMA.Encapsulate(quicMA), nil
+	switch version {
+	case quic.VersionDraft29:
+		return udpMA.Encapsulate(quicDraft29MA), nil
+	case quic.Version1:
+		return udpMA.Encapsulate(quicV1MA), nil
+	default:
+		return nil, errors.New("unknown quic version")
+	}
 }
 
-func fromQuicMultiaddr(addr ma.Multiaddr) (net.Addr, error) {
-	return manet.ToNetAddr(addr.Decapsulate(quicMA))
+func fromQuicMultiaddr(addr ma.Multiaddr) (net.Addr, quic.VersionNumber, error) {
```

Comment:
Slight preference for keeping this such that we iterate the multiaddr fewer times. (I've even made a change to join the multiaddrs once at the end instead of each step).

Reply:
In general, it feels "wrong" to do that much multiaddr magic in libp2p packages. Maybe the API we expose in `multiaddr` is insufficient? Or are we over-optimizing things here?

Anyway, no strong feelings, I'm ok with keeping this as is, just something to think about.

Reply:
> Maybe the API we expose in multiaddr is insufficient

I think so. I think there are some things we can add that would replace 80% of our uses of `ForEach`.